cmd/utils.go

```diff
@@ -318,3 +325,89 @@ func isURLContains(srcURL, tgtURL, sep string) bool {
 	}
 	return false
 }
+
+// Return the file system attribute as string; containing mode,
+// uid, gid, uname, Gname, atime, mtime, ctime and md5
+func getFileSystemAttrs(file string) (string, error) {
+	var uname, gname, atime, mtime, ctime, mode, uid, gid string
+	fi, err := os.Stat(file)
+	if err != nil {
+		return "", err
+	}
+	st := fi.Sys().(*syscall.Stat_t)
+	mode = fmt.Sprint(st.Mode)
+	uid = fmt.Sprint(st.Uid)
+	gid = fmt.Sprint(st.Gid)
+
+	i, err := user.LookupId(strconv.FormatUint(uint64(st.Uid), 10))
+	if err != nil {
+		return "", err
+	}
+	uname = i.Username
+
+	g, err := user.LookupGroupId(strconv.FormatUint(uint64(st.Gid), 10))
+	if err != nil {
+		return "", err
+	}
+	gname = g.Name
+
+	t, err := times.Stat(file)
+	if err != nil {
+		return "", err
+	}
+	atime = fmt.Sprint(t.AccessTime().Unix())
+	mtime = fmt.Sprint(t.ModTime().Unix())
+
+	if t.HasChangeTime() {
+		ctime = fmt.Sprint(t.ChangeTime().Unix())
+	}
+
+	md5, err := hashFilemd5(file)
+	if err != nil {
+		return "", err
+	}
+	fileAttr := fmt.Sprintf("atime:%s/ctime:%s/gid:%s/gname:%s/md5:%s/mode:%s/mtime:%s/uid:%s/uname:%s", atime, ctime, gid, gname, md5, mode, mtime, uid, uname)
+
+	return fileAttr, nil
+}
+
+// Return the md5 of the file as string
+func hashFilemd5(filePath string) (string, error) {
+	//Initialize variable returnMD5String now in case an error has to be returned
+	var returnMD5String string
+
+	//Open the passed argument and check for any error
+	file, err := os.Open(filePath)
+	if err != nil {
+		return returnMD5String, err
+	}
+
+	//Tell the program to call the following function when the current function returns
+	defer file.Close()
+
+	//Open a new hash interface to write to
+	hash := md5.New()
+
+	//Copy the file in the hash interface and check for any error
+	if _, err := io.Copy(hash, file); err != nil {
+		return returnMD5String, err
+	}
+
+	//Get the 16 bytes hash
+	hashInBytes := hash.Sum(nil)[:16]
+
+	//Convert the bytes to a string
+	returnMD5String = hex.EncodeToString(hashInBytes)
+
+	return returnMD5String, nil
+}
+
+// Returns a map by parsing the value of X-Amz-Meta-Mc-Attrs/X-Amz-Meta-s3Cmd-Attrs
+func parseAttribute(attrs string) map[string]string {
+	attribute := make(map[string]string)
+	for _, attr := range strings.Split(attrs, "/") {
+		param := strings.Split(attr, ":")
+		attribute[param[0]] = param[1]
+	}
+	return attribute
```

Comment:
This code panics with an empty string as input, write tests to ensure that all negative cases are handled. 

https://play.golang.org/p/RWGfz35f-18 - see all the offending inputs

Reply:
> This code panics with an empty string as input, write tests to ensure that all negative cases are handled.
> 
> https://play.golang.org/p/RWGfz35f-18 - see all the offending inputs

@harshavardhana  Should i add `recover()` in the `parseAttribute` function. How to add the test case for recover. Not clear on this, need your input.

Reply:
No, I am asking to write this part of the code such that there is no panic and we recover appropriately - we shouldn't add `recover()`